ppl/archive/immcache/immcache.go

```diff
@@ -5,29 +5,98 @@ package immcache
 import (
 	"context"
 	"flag"
+	"fmt"
+	"time"
 
+	"github.com/brimsec/zq/cli"
 	"github.com/brimsec/zq/pkg/iosrc"
+	"github.com/go-redis/redis/v8"
 	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
 )
 
+type Kind string
+
+const (
+	KindNone  Kind = "none"
+	KindLocal Kind = "local"
+	KindRedis Kind = "redis"
+)
+
+func (k *Kind) Set(s string) error {
+	switch s {
+	case "none", "":
+		*k = KindNone
+	case "local":
+		*k = KindLocal
+	case "redis":
+		*k = KindRedis
+	default:
+		return fmt.Errorf("unknown immutable cache kind: %q", s)
+	}
+	return nil
+}
+
+func (k Kind) String() string {
+	return string(k)
+}
+
 type ImmutableCache interface {
 	ReadFile(context.Context, iosrc.URI) ([]byte, error)
 }
 
 type Config struct {
+	Kind Kind
 	// LocalCacheSize specifies the number of immutable files to keep in a
 	// local lru cache used to speed up searches. Values less than or equal to 0
 	// (default), disables local caching of immutable files.
 	LocalCacheSize int
+	// RedisKeyExpiration is the time redis will keep the key of the immutable
+	// file around before deleting this. If RedisKeyExpiration == 0 the key will
+	// be kept around indefinitely. Because this is supposed to be used as an
+	// LRU cache, if the redis server eviction policy is set volitale-lru the
+	// value of RedisKeyExpiration should not be set to 0.
+	RedisKeyExpiration cli.Duration
 }
 
 func (c *Config) SetFlags(fs *flag.FlagSet) {
-	fs.IntVar(&c.LocalCacheSize, "immcache.localsize", 0, "enables local caching of up to N small files (disabled if 0)")
+	c.RedisKeyExpiration = cli.Duration(time.Hour * 24)
+	fs.Var(&c.Kind, "immcache.kind", "kind of immutable cache")
+	fs.IntVar(&c.LocalCacheSize, "immcache.local.size", 128, "number of small files to keep in local cache")
+	fs.Var(&c.RedisKeyExpiration, "immcache.redis.keyexpiry", "expiration duration of immutable keys")
```

Comment:
There is a `fs.DurationVar`, is there a reason it wouldn't work here?

Reply:
I did not know that!